pkg/framework/machinesets.go

```diff
@@ -369,6 +449,106 @@ func WaitForMachineSetDelete(c runtimeclient.Client, machineSet *machinev1.Machi
 	WaitForMachineSetsDeleted(c, machineSet)
 }
 
+// WaitForSpotMachineSet waits for all Machines belonging to the machineSet to be running and their nodes to be ready.
+// Unlike WaitForMachineSet, this function does not fail the test when machine cannoct be provisioned due to insufficient spot capacity.
+func WaitForSpotMachineSet(c client.Client, name string) error {
+	machineSet, err := GetMachineSet(c, name)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Retry until the MachineSet is ready.
+	err = wait.PollImmediate(RetryMedium, WaitLong, func() (bool, error) {
+		machines, err := GetMachinesFromMachineSet(c, machineSet)
+		if err != nil {
+			return true, err
+		}
+
+		replicas := pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0)
+		if len(machines) != int(replicas) {
+			klog.Infof("%q: found %d Machines, but MachineSet has %d replicas", name, len(machines), int(replicas))
+			return false, nil
+		}
+
+		failed := FilterMachines(machines, MachinePhaseFailed)
+		if len(failed) > 0 {
+			// if there are failed machines, print them out before we exit
+			klog.Errorf("found %d Machines in failed phase: ", len(failed))
+			for _, m := range failed {
+				reason := "failureReason not present in Machine.status"
+				if m.Status.ErrorReason != nil {
+					reason = string(*m.Status.ErrorReason)
+				}
+				message := "failureMessage not present in Machine.status"
+				if m.Status.ErrorMessage != nil {
+					message = string(*m.Status.ErrorMessage)
+				}
+				klog.Errorf("Failed machine: %s, Reason: %s, Message: %s", m.Name, reason, message)
+			}
+		}
+		Expect(len(failed)).To(Equal(0))
```

Comment:
Probably we shouldn't call ginkgo functions here directly - just return an error if len(failed) > 0